pkg/daemon/server/service/pipeline_metrics_query_service.go

```diff
@@ -0,0 +1,192 @@
+		bufferLength, bufferUsageLimit := getBufferLimits(ps.pipeline, edge)
+		usage := float64(bufferInfo.TotalMessages) / float64(bufferLength)
+		if x := (float64(bufferInfo.PendingCount) + float64(bufferInfo.AckPendingCount)) / float64(bufferLength); x < usage {
+			usage = x
+		}
+		b := &daemon.BufferInfo{
+			Pipeline:         &ps.pipeline.Name,
+			FromVertex:       pointer.String(fmt.Sprintf("%v", edge.From)),
+			ToVertex:         pointer.String(fmt.Sprintf("%v", edge.To)),
+			BufferName:       pointer.String(fmt.Sprintf("%v", buffer)),
+			PendingCount:     &bufferInfo.PendingCount,
+			AckPendingCount:  &bufferInfo.AckPendingCount,
+			TotalMessages:    &bufferInfo.TotalMessages,
+			BufferLength:     &bufferLength,
+			BufferUsageLimit: &bufferUsageLimit,
+			BufferUsage:      &usage,
+			IsFull:           pointer.Bool(usage >= bufferUsageLimit),
+		}
+		buffers = append(buffers, b)
+	}
+	resp.Buffers = buffers
+	return resp, nil
+}
+
+// GetBuffer is used to obtain one buffer information of a pipeline
+func (ps *pipelineMetricsQueryService) GetBuffer(ctx context.Context, req *daemon.GetBufferRequest) (*daemon.GetBufferResponse, error) {
+	bufferInfo, err := ps.isbsvcClient.GetBufferInfo(ctx, v1alpha1.Buffer{Name: *req.Buffer, Type: v1alpha1.EdgeBuffer})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get information of buffer %q:%v", *req.Buffer, err)
+	}
+	edge := ps.pipeline.FindEdgeWithBuffer(*req.Buffer)
+	if edge == nil {
+		return nil, fmt.Errorf("unexpected error, buffer %q not found from the pipeline", *req.Buffer)
+	}
+	bufferLength, bufferUsageLimit := getBufferLimits(ps.pipeline, *edge)
+	usage := float64(bufferInfo.TotalMessages) / float64(bufferLength)
+	if x := (float64(bufferInfo.PendingCount) + float64(bufferInfo.AckPendingCount)) / float64(bufferLength); x < usage {
+		usage = x
+	}
+	b := &daemon.BufferInfo{
+		Pipeline:         &ps.pipeline.Name,
+		FromVertex:       &edge.From,
+		ToVertex:         &edge.To,
+		BufferName:       req.Buffer,
+		PendingCount:     &bufferInfo.PendingCount,
+		AckPendingCount:  &bufferInfo.AckPendingCount,
+		TotalMessages:    &bufferInfo.TotalMessages,
+		BufferLength:     &bufferLength,
+		BufferUsageLimit: &bufferUsageLimit,
+		BufferUsage:      &usage,
+		IsFull:           pointer.Bool(usage >= bufferUsageLimit),
+	}
+	resp := new(daemon.GetBufferResponse)
+	resp.Buffer = b
+	return resp, nil
+}
+
+// GetVertexMetrics is used to query the metrics service and is used to obtain the processing rate of a given vertex for 1m, 5m and 15m.
+// In the future maybe latency will also be added here?
+// Should this method live here or maybe another file?
+func (ps *pipelineMetricsQueryService) GetVertexMetrics(ctx context.Context, req *daemon.GetVertexMetricsRequest) (*daemon.GetVertexMetricsResponse, error) {
+	log := logging.FromContext(ctx)
+	resp := new(daemon.GetVertexMetricsResponse)
+
+	metricsPort := strconv.Itoa(v1alpha1.VertexMetricsPort)
```

Comment:
nit: it can be used below in line 131 with `fmt.Sprintf("xxxxxx%v", v1alpha1.VertexMetricsPort)` without converting to string.